회원가입 및 로그인, 아카이브 CRUD

NavyHubb opened the pull request:
### 변경사항
**AS-IS**
프로젝트 초기 세팅

**TO-BE**
- 회원가입 및 로그인
- 아카이브 생성, 수정, 삭제
- 아카이브 단건 조회

각 API 기능을 수행하기에 앞서 필요한 사용자 권한 인가 과정은 추후에 구현할 예정입니다.
### 테스트
<!-- 본 변경사항이 테스트가 되었는지 기술해주세요 --> 
- [ ] 테스트 코드
- 회원가입
- [x] API 테스트

NavyHubb committed:
feat: 초기 엔티티 세팅

NavyHubb committed:
feat: 회원가입 기능 구현

- 비밀번호는 암호화하여 DB에 저장 (BCryptPasswordEncoder 사용)

NavyHubb committed:
feat: Swagger 설정

- 의존성 및 Config 파일 추가

NavyHubb committed:
feat: 로그인 기능 구현

- 로그인 시 이메일과 비밀번호를 입력받는다
- 정보가 일치하면 JWT를 반환한다

NavyHubb committed:
test: 회원가입 테스트 코드 작성

NavyHubb committed:
docs: 리드미 수정

NavyHubb committed:
feat: Naver API로부터의 도서 정보 조회 기능 구현

NavyHubb committed:
feat: 아카이브 생성, 수정, 삭제 기능 구현

- 생성: 기존에 같은 도서명을 가진 아카이브가 있을 시 생성 불가
- 삭제: 아카이브와 연관된 리뷰와 스크랩들의 삭제 연동
기능은 추후 구현 예정

NavyHubb committed:
style: 메서드 명 수정

as-is:
search

to-be:
searchList

NavyHubb committed:
docs: 리드미 수정

- ERD 수정

NavyHubb committed:
feat: 아카이브 상세 조회

NavyHubb committed:
docs: 리드미 수정

- 기능 추가
- ERD 수정

NavyHubb commented:
첫 PR인지라 여러 커밋이 뭉쳤네요..! 추후 PR은 기능단위로 분할하여 올리겠습니다.

yuuzinn commented:
코드 잘 봤습니다. 고생하셨습니다..!!!! 👍

mingming-mentor commented on src/main/java/com/green/greenbook/controller/ArchiveController.java:
```diff
@@ -0,0 +1,44 @@
+    private final ArchiveService archiveService;
+
+    @PostMapping
+    public ResponseEntity<Archive> create(@RequestBody ArchiveForm form) {
```
다른 스터디원분 코드에서 드렸던 리뷰인데 동일하게 드려볼게요 

--------
service에서 사용하는 dto와 controller에서 받아오는 request는 분리하는 것이 좋습니다. 지금은 controller에서 하나의 service만 호출하긴 하지만 여러개의 service 로직을 호출할때도 있기 때문에 controller단과 service를 분리해주시면 좋아요. 참고하실만한 블로그 링크도 같이 드려볼게요  
https://techblog.woowahan.com/2711/ 의 `2. Controller와 Service 레이어의 강한 결합` 참고해주세요

NavyHubb replied:
감사합니다. 참고해서 반영해볼게요!

NavyHubb replied:
@mingming-mentor '서비스로 입력되는 dto'와 '서비스가 반환하는 dto'도 서로 구분하는 게 유지보수 측면에서 효율적일까요?

mingming-mentor replied:
@NavyHubb 넵 구분하는 것이 더 효율적일듯 합니다. 아카이브를 생성할때 필요한 데이터와, 아카이브 생성로직을 수행했을때 결과로 반환해야하는 데이터는 다를 수가 있으니깐요. 
코드를 여러명이서 개발하고 오래되고 복잡해지면, 어떤 데이터가 입력에 필요한 값인지, 어떤 데이터가 응답에 필요한 값인지 명확히 구분하기가 어려워지기도 합니다. 그래서 저는 분리해서 개발하는 편을 선호합니다

NavyHubb replied:
@mingming-mentor 
각 쓰임새에 따라 DTO를 마련해야한다는 건 이해가 되는데 각각을 네이밍하는 게 고민이 되네요. 다들 각자의 방식으로 하고 있는 듯한데 혹시 선호하시는 네이밍 컨벤션이 있을까요?

mingming-mentor replied:
@NavyHubb 님
저는 개인적으로는 controller -> service로 데이터를 넘겨줄때는 객체화해서 넘겨주지 않고 param으로 다 빼서 넣어주고 있습니다. 
즉 현재와 같은 경우 

`archiveService.create(form)
`를
`archiveService.create(form.getTitle(), form.getAuthor(), form.getPublisher())
`형태로 많이 사용하는 것 같아요.

service에서 반환하는 객체들은 `xxDto`라는 이름을 사용해주는 편이구요
controller에서 input/return 하는 객체들은 각각 `xxxRequest`, `xxxResponse` 라는 이름으로 사용해주고 있습니다. 
다만 이 부분은 개인취향인 부분이므로 원하시는 네이밍을 사용하셔도 무방할것 같아요.

mingming-mentor commented on src/main/java/com/green/greenbook/controller/BookController.java:
```diff
@@ -0,0 +1,63 @@
+    @GetMapping("/book")
+    public List<BookVO> searchList(@RequestBody SearchForm form) {
+
+        String clientId = "o2E_K7nOKatROhYYz1RA";
```
서비스 로직이 controller에 들어가 있는 걸로 보여요. 이 부분을 service를 따로 나눠서 분리해주시면 좋을 것 같아요. 
그리고 naver api로부터 데이터를 가지고 오는 부분은 service 로직이라기보단 component로직으로 분리해주시면 좋을 것 같습니다. 즉 
controller -> search service -> naver api component 로 나눠주심 좋을 것 같아요. 
각각의 역할은 
controller : 요청을 받고 service에서 반환한 값을 view에 맞춰서 변환
service : 검색하고자 하는 param을 서비스 로직에 의해 변경하거나, 권한 체크-validation 등 요구사항에 맞는 서비스의 로직 수행 (현재 로직에서 권한체크나 validation은 없지만 이런 부분이 service에 보통 들어가는 부분이라 기재합니다)
component : 외부와의 통신 등..

mingming-mentor commented on src/main/java/com/green/greenbook/controller/BookController.java:
```diff
@@ -0,0 +1,63 @@
+            .header("X-Naver-Client-Id", clientId)
+            .header("X-Naver-Client-Secret", clientSecret)
+            .build();
+        RestTemplate restTemplate = new RestTemplate();
```
RestTemplate은 생성하는데 비용이 크기 때문에 bean으로 만들어서 관리해주세요

mingming-mentor replied:
요 bean을 만들때 naver api용으로 만들면 header부분에 넣어주는 데이터를 아예 셋팅해서 넣어줘도 될것 같아요

mingming-mentor commented on src/main/java/com/green/greenbook/controller/BookController.java:
```diff
@@ -0,0 +1,63 @@
+        RestTemplate restTemplate = new RestTemplate();
+        ResponseEntity<String> resp = restTemplate.exchange(req, String.class);
+
+        ObjectMapper om = new ObjectMapper();
```
object mapper도 생성하는데 비용이 크기 때문에 bean으로 만들어서 관리해주세요

mingming-mentor commented on src/main/java/com/green/greenbook/controller/BookController.java:
```diff
@@ -0,0 +1,63 @@
+        try {
+            resultVO = om.readValue(resp.getBody(), NaverResultVO.class);
+        } catch (JsonMappingException e) {
+            e.printStackTrace();
```
printStacTrace로 찍기만한다면 api 응답받는 쪽은 응답을 못받을거에요. log로 찍고 e는 던져서 ExceptionController 등에서 일괄처리하여 응답을 내려주는 방식으로 개발하면 어떨까요?

mingming-mentor commented on src/main/java/com/green/greenbook/controller/SignUpController.java:
```diff
@@ -0,0 +1,42 @@
+
+    @PostMapping
+    public ResponseEntity<Object> signUp(@RequestBody @Valid SignUpForm form, Errors errors) {
+        if (errors.hasErrors()) {
```
요 Errors는 어떻게 받아오게 되나요? @ModelAttribute에서 사용하는 방식과 혼합된거같아서 여쭈어보아요.

mingming-mentor commented on src/main/java/com/green/greenbook/domain/model/Archive.java:
```diff
@@ -0,0 +1,52 @@
+    private List<Review> reviewList =
+            new ArrayList<>();
+
+    private String title;
```
아카이브의 title은 ERD에 없어보이는데요, 확인 부탁드릴게요~

mingming-mentor replied:
구독수는 ERD에 있어보이는데 이 부분은 따로 없네요. 이 부분도 확인해주세요

NavyHubb replied:
오잉 아카이브 테이블에 title(도서명) 있지 않나요?😯

mingming-mentor replied:
@NavyHubb 도서명이 타이틀이군요. 확인했습니다! 잘못봐서 죄송해요ㅠ

mingming-mentor commented on src/main/java/com/green/greenbook/domain/model/Member.java:
```diff
@@ -0,0 +1,53 @@
+
+    @Builder.Default
+    @OneToMany(mappedBy = "member")
+    private List<Scrap> scraps
```
요 관계는 member 에 꼭 있어야하나용? scrap api에서 user 아이디를 이용해 별도로 호출한다면 굳이 JPA에서 선언해줄 필요는 없어서 여쭤봅니다

NavyHubb replied:
회원 개인 페이지에서 본인이 스크랩한 목록을 띄우려는 목적으로 만들어두었는데 말씀해주신대로 한다면 굳이 있어야할까 싶네요!

mingming-mentor commented on src/main/java/com/green/greenbook/domain/model/Review.java:
```diff
@@ -0,0 +1,37 @@
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+public class Review extends BaseEntity {
```
삭제일이 빠져있네요

mingming-mentor commented on src/main/java/com/green/greenbook/service/ArchiveService.java:
```diff
@@ -0,0 +1,60 @@
+
+    public Archive create(ArchiveForm form) {
+        if (archiveRepository.findByTitle(form.getTitle()).isPresent()) {
+            throw new CustomException(ErrorCode.ALREADY_REGISTERED_BOOKTITLE);
```
제목이 같은 책은 1권뿐인지가 궁금하네요. 예를 들어 <탈무드> 같은 경우는 여러 출판사에서 탈무드라는 이름으로 책을 내서 여러권이 있지 않나요?

mingming-mentor replied:
title이 아니라 책 id같은걸로 해야하지 않나 싶어요. naver에서 관리하는 책마다 발급되는 id가 없을까요?

NavyHubb replied:
도서 고유 번호인 ISBN 값이면 적당할듯 싶어요! 이 값을 담을 컬럼을 Archive 클래스에 추가하고, 이 컬럼을 기준으로 중복 검사를 하는 식으로 해보겠습니다.

mingming-mentor commented on src/main/java/com/green/greenbook/service/ArchiveService.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    public Archive update(Long archiveId, ArchiveForm form) {
+        Archive archive = getArchive(archiveId);
```
요건 어떤 유저가 호출할수 있나요? 유저가 호출이 가능한 부분이라면 권한 체크는 없어도 될지 궁금해요

NavyHubb replied:
정책적인 문제일듯 싶은데 좀더 고민해보겠습니다. 권한 체크 부분은 API 테스트의 편의상 추후에 구현하려고 하는데 괜찮을까요?

KimHyoJin replied:
@NavyHubb 넵 문제없습니다!

mingming-mentor replied:
@NavyHubb 넵 문제없습니디다

mingming-mentor commented on src/main/java/com/green/greenbook/service/SignUpService.java:
```diff
@@ -0,0 +1,50 @@
+    private Member fromForm(SignUpForm form) {
+        return Member.builder()
+            .email(form.getEmail())
+            .password(new BCryptPasswordEncoder().encode(form.getPassword()))
```
BCryptPasswordEncoder 도 bean으로 관리하시면 좋을듯해요

mingming-mentor commented on src/main/java/com/green/greenbook/util/Aes256Util.java:
```diff
@@ -0,0 +1,49 @@
+
+            return Base64.encodeBase64String(encrypted);
+        } catch (Exception e) {
+            return null;
```
error log등을 찍어주지 않아도 될까요?

NavyHubb replied:
로그 처리 반영해보겠습니다.

mingming-mentor reviewed (approved):
리뷰드렸습니다. ERD는 전체적으로 작성하신 코드와 README 가 동일한지 체크부탁드려요. 다음에는 말씀주신대로 코드가 조금 쪼개져서 올라오면 좋을듯하네요. 수고 많으셨습니다

NavyHubb committed:
feat: 논리 삭제 적용

- Archive, Reivew 클래스에 대한 @SQLDelete, @Where 조건 적용
- deletedAt 컬럼 추가
- 리드미 업데이트
